src/rendezvous/rendezvous.ts

```diff
@@ -0,0 +1,228 @@
+        }
+        const { outcome, device_id: deviceId, device_key: deviceKey } = res;
+
+        if (outcome !== 'success') {
+            throw new Error('Linking failed');
+        }
+
+        this.newDeviceId = deviceId;
+        this.newDeviceKey = deviceKey;
+
+        return deviceId;
+    }
+
+    private async verifyAndCrossSignDevice(deviceInfo: DeviceInfo) {
+        if (!this.client.crypto) {
+            throw new Error('Crypto not available on client');
+        }
+
+        if (!this.newDeviceId) {
+            throw new Error('No new device ID set');
+        }
+
+        // check that keys received from the server for the new device match those received from the device itself
+        if (deviceInfo.getFingerprint() !== this.newDeviceKey) {
+            throw new Error(
+                `New device has different keys than expected: ${this.newDeviceKey} vs ${deviceInfo.getFingerprint()}`,
+            );
+        }
+
+        const userId = this.client.getUserId();
+
+        if (!userId) {
+            throw new Error('No user ID set');
+        }
+        // mark the device as verified locally + cross sign
+        logger.info(`Marking device ${this.newDeviceId} as verified`);
+        const info = await this.client.crypto.setDeviceVerification(
+            userId,
+            this.newDeviceId,
+            true, false, true,
+        );
+
+        const masterPublicKey = this.client.crypto.crossSigningInfo.getId('master');
+
+        await this.send({
+            type: PayloadType.Finish,
+            outcome: 'verified',
+            verifying_device_id: this.client.getDeviceId(),
+            verifying_device_key: this.client.getDeviceEd25519Key(),
+            master_key: masterPublicKey,
+        });
+
+        return info;
+    }
+
+    async verifyNewDeviceOnExistingDevice(timeout = 10 * 1000): Promise<DeviceInfo | CrossSigningInfo | undefined> {
+        if (!this.newDeviceId) {
+            throw new Error('No new device to sign');
+        }
+
+        if (!this.newDeviceKey) {
+            logger.info("No new device key to sign");
+            return undefined;
+        }
+
+        if (!this.client.crypto) {
+            throw new Error('Crypto not available on client');
+        }
+
+        const userId = this.client.getUserId();
+
+        if (!userId) {
+            throw new Error('No user ID set');
+        }
+
+        {
+            const deviceInfo = this.client.crypto.getStoredDevice(userId, this.newDeviceId);
+
+            if (deviceInfo) {
+                return await this.verifyAndCrossSignDevice(deviceInfo);
+            }
+        }
+
+        logger.info("Going to wait for new device to be online");
+        await sleep(timeout);
+
+        {
+            const deviceInfo = this.client.crypto.getStoredDevice(userId, this.newDeviceId);
+
+            if (deviceInfo) {
+                return await this.verifyAndCrossSignDevice(deviceInfo);
+            }
+        }
```

Comment:
another random scope?

Reply:
It's because there was a variable name re-use. I've refactored to be more elegant.